shelley/chain-and-ledger/executable-spec/test/Generator/Core.hs

```diff
@@ -0,0 +1,97 @@
+import           Coin (Coin (..))
+import           Keys (pattern KeyPair, pattern SKey, pattern VKey, hashKey, vKey)
+import           LedgerState (genesisCoins)
+import           MockTypes (Addr, Credential, KeyPair, KeyPairs, SKey, TxOut, UTxO, VKey)
+import           Tx (pattern TxOut)
+import           TxData (pattern AddrBase, pattern KeyHashObj)
+
+mkKeyTuple :: (Word64, Word64, Word64, Word64, Word64) -> (VKey, SKey)
+mkKeyTuple seed = fst . withDRG (drgNewTest seed) $ do
+  sk <- genKeyDSIGN
+  return ( VKey $ deriveVerKeyDSIGN sk
+         , SKey sk)
+
+mkKeyPairs :: Word64 -> (KeyPair, KeyPair)
+mkKeyPairs n
+  = (mkKeyPair (2*n), mkKeyPair (2*n+1))
+  where
+    mkKeyPair n_ = uncurry KeyPair (mkKeyTuple (n_,n_,n_,n_,n_))
+
+-- | Constant list of KeyPairs intended to be used in the generators.
+traceKeyPairs :: KeyPairs
+traceKeyPairs = mkKeyPairs <$> [1 .. 50]
+
+-- | Select between _lower_ and _upper_ keys from 'traceKeyPairs'
+someKeyPairs :: Int -> Int -> Gen KeyPairs
+someKeyPairs lower upper =
+  take
+    <$> Gen.integral (Range.linear lower upper)
+    <*> Gen.shuffle traceKeyPairs
+
+-- | Find first matching key pair for address. Returns the matching key pair
+-- where the first element of the pair matched the hash in 'addr'.
+findPayKeyPair :: Addr -> KeyPairs -> KeyPair
+findPayKeyPair (AddrBase (KeyHashObj addr) _) keyList =
+    case matches of
+      []    -> error "findPayKeyPair: could not find a match for the given address"
+      (x:_) -> fst x
+    where
+      matches = filter (\(pay, _) -> addr == hashKey (vKey pay)) keyList
+findPayKeyPair _ _ = error "findPayKeyPair: expects only AddrBase addresses"
+
+-- | Select one random verification staking key from list of pairs of KeyPair.
+pickStakeKey :: KeyPairs -> Gen VKey
+pickStakeKey keys = vKey . snd <$> Gen.element keys
+
+toAddr :: (KeyPair, KeyPair) -> Addr
+toAddr (payKey, stakeKey) = AddrBase (toCred payKey) (toCred stakeKey)
+
+toCred :: KeyPair -> MockTypes.Credential
+toCred k = KeyHashObj . hashKey $ vKey k
+
+-- | Generates a list of coins for the given 'Addr' and produced a 'TxOut' for each 'Addr'
+genTxOut :: [Addr] -> Gen [TxOut]
+genTxOut addrs = do
+  ys <- genCoinList 100 10000 (length addrs) (length addrs)
+  return (uncurry TxOut <$> zip addrs ys)
+
+-- | Generates a list of 'Coin' values of length between 'lower' and 'upper'
+-- and with values between 'minCoin' and 'maxCoin'.
+genCoinList :: Integer -> Integer -> Int -> Int -> Gen [Coin]
+genCoinList minCoin maxCoin lower upper = do
+  xs <- Gen.list (Range.linear lower upper)
+        $ Gen.integral (Range.exponential minCoin maxCoin)
+  return (Coin <$> xs)
+
+genUtxo0 :: KeyPairs -> Gen UTxO
+genUtxo0 keys = do
+  let addrs = fmap toAddr keys
+  outs <- genTxOut addrs
+  return (genesisCoins outs)
+
+unsafeMkUnitInterval :: Rational -> UnitInterval
+unsafeMkUnitInterval r =
+  fromMaybe (error "could not construct unit interval") $ mkUnitInterval r
```

Comment:
this too could be imported from Examples ?

Reply:
👍